rfcs/63-nursery.md

```diff
@@ -0,0 +1,172 @@
+
+Very small or uncontroversial crates may be suitable for a direct PR via the ordinary mechanism.
+
+### Setting up a nursery crate
+
+Once accepted, the repository will be transferred to the Bevy organization under the same name and `crates.io` identity.
+
+The existing crate maintainers will have complete admin permissions over the new crate,
+and any other Bevy maintainers will be granted write permissions (or the `bors` equivalent).
+
+A new milestone will be created, called `upstreaming`, which will track the work that needs to be completed before the crate can be submitted as a PR to `bevy` itself.
+
+Finally, the issues and PRs from that crate will be added to the `#bevy-nursery` channel on Discord to improve discoverability.
+
+### Preparing a crate to be upstreamed
+
+Before a nursery crate can be accepted to Bevy itself, the following conditions must be met:
+
+1. All items in the `upstreaming` milestone must be complete.
+2. A final release must be published, encompassing all PRs merged to date.
+3. The Bevy maintainers must collectively agree to accept the crate upstream in its current form.
+
+This checklist may be helpful when considering issues for the `upstreaming` milestone:
+
+- [] are the examples and documentation helpful and complete?
+- [] do all features work as intended?
+- [] is the code quality acceptable?
+- [] is there any urgent highly-experimental work that should be completed before attempting to integrate more closely with the rest of the engine?
+
+Crates **do not have to be complete** to be merged into Bevy itself: merely polished and clearly useful.
+
+The PR to merge the crate into Bevy itself should be very simple, and contain only the bare minimum change set required to enable interoperability.
+Further integration work should be done in follow-up PRs.
+If you are an interested community member, raise issues and submit fixes on the nursery repository **before** the crate is being upstreamed.
+
+Once the upstreaming PR is merged, the nursery crate's repo will be archived and any still-relevant issues will be migrated to the main Bevy repo.
+
+### Maintaining upstreamed crates
+
+When a nursery crate is upstreamed into the main engine,
+the core maintainer team of the nursery crate will be offered (but do not need to accept)
+**scoped merge rights** for the Bevy project.
+
+Nursery crate maintainers have shown themselves to have the judgement and reliability needed to manage part of the core engine:
+upstreaming a crate should not impose any additional barriers to the crate's development and maintenance.
+Following the standard rules for community reviews, they will be able to merge PRs that affect their nursery crate and directly related areas of the engine.
+Controversial PRs in their area will require either the sign-off of the project lead *or* the nursery crate maintainer, although like always, consulting experts is always wise!
```

Comment:
I strongly disagree with backdooring our decision making process here, especially as the default course of action. Controversial decisions are often controversial because they have implications in a wider context. While in some cases nursery crate owners _may_ have proven their ability to make consistent / generally likeable decisions, when slotted into a wider context of Bevy the correct choice can change. Someone could make a "perfect" crate that is beloved, consistent, fully featured, well documented, etc and it could _still_ be a terrible choice for Bevy. Nursery crate authors have not proven themselves in that context.

Consider this example (which I expect will be common): pretend we add a crate to the nursery and then as part of the nursery process, we significantly alter the design to fit Bevy's needs. If the "vision" required substantial alternations due to feedback from maintainers, can we really trust the author to make the right calls going forward?

Reply:
Now that you mention it, I'm uncomfortable giving scoped merge rights to nursery owners *at all*. I don't think it should function any different from a PR - it's not as if we give those contributors merge rights to the area of the code they added.

Reply:
It could become a problem though if a crate gets merged upstream and then stagnates in development due to stricter and slower processes.
Then someone might create a new crate to solve this and we get ecosystem duplication again.

Not sure if this _will_ happen, but definitely something to consider IMO.

Reply:
Imo the slower process (and the ability for the 3rd party ecosystem to out compete it in development velocity) is a feature, not a bug. If we are making controversial decisions at the same rate as someone unencumbered by Bevy quality / consistency standards, that would be concerning.